### DISCUSSION TITLE
Using one marimo installation with multiple environments 🐶

### DISCUSSION THREAD

**Discussion by liquidcarbon** [marimo-team/marimo#2994] [Show and tell]
## marimo + puppy = 💚

This is a recipe for structuring your marimo projects for multiple environments using a package manager I developed called [puppy](https://github.com/liquidcarbon/puppy).

The same approach will work for Jupyter.

Puppy is a CLI wrapper for [pixi](https://github.com/prefix-dev/pixi/) and [uv](https://github.com/astral-sh/uv).

The TLDR of puppy:
1) create a folder, install python + whatever toys you will use across all your projects, e.g. uv, marimo, jupyter, pytest, etc. (also conda and non-python stuff)
2) use `pup add` commands to build virtual environments defined by `pyproject.toml` files with uv
3) use `pup.fetch()` to "activate" the venvs from any interactive shell (python, marimo, ipython, jupyter)

## Let's build

Suppose you're trying out some cool new [anywidget](https://docs.marimo.io/guides/integrating_with_marimo/custom_ui_plugins.html) tools that you plan to use with both marimo and jupyter.

1. Start in an empty folder

```bash
mkdir puppy-marimo && cd puppy-marimo
```

2. Install puppy
```bash
curl -fsSL https://raw.githubusercontent.com/liquidcarbon/puppy/main/pup.sh | bash
```

output:
```
✨ pixi 0.35.0 found
✔ Created /home/a/puppy-marimo/pixi.toml
✔ Added python=3.12
✔ Added uv>=0
🟣 uv 0.5.4
✔ Added click>=8
```

3.  Install marimo, jupyter, and anything else that you'll use from root environment.
```bash
pixi add jupyter marimo quarto
```

output:
```
✔ Added jupyter >=1.1.1,<2
✔ Added marimo >=0.9.26,<0.10
✔ Added quarto >=1.5.57,<2
```

4. Create new virtual environment called `newenv` and install [tldraw](https://github.com/kolibril13/jupyter-tldraw):
```bash
pup add newenv tldraw
```

output:

```
[2024-11-27 15:18:05] 🐶 heard: pup new newenv
[2024-11-27 15:18:05] 🐶 said: pixi run uv init /home/a/puppy-marimo/newenv -p /home/a/puppy-marimo/.pixi/envs/default/bin/python --no-workspace
Initialized project `newenv` at `/home/a/puppy-marimo/newenv`
[2024-11-27 15:18:05] 🐶 said: pixi run uv venv /home/a/puppy-marimo/newenv/.venv -p /home/a/puppy-marimo/.pixi/envs/default/bin/python
Using CPython 3.12.7 interpreter at: .pixi/envs/default/bin/python
Creating virtual environment at: newenv/.venv
Activate with: source newenv/.venv/bin/activate
[2024-11-27 15:18:06] 🐶 heard: pup add newenv tldraw
[2024-11-27 15:18:06] 🐶 said: pixi run uv add tldraw --project /home/a/puppy-marimo/newenv
Resolved 47 packages in 31ms
Installed 41 packages in 45ms
 + anywidget==0.9.13
 + asttokens==2.4.1
...
```

5. Instead of installing a whole new marimo into each new environment, place a copy of marimo's executable into `newenv`, patching it with [pup.fetch](https://github.com/liquidcarbon/puppy?tab=readme-ov-file#using-pup-as-a-module) 

```bash
VENV=newenv
head -5 .pixi/envs/default/bin/marimo > "$VENV/.venv/bin/marimo"
echo -e '\nimport pup\npup.fetch("'$VENV'")\n' >> "$VENV/.venv/bin/marimo"
tail -5 .pixi/envs/default/bin/marimo >> "$VENV/.venv/bin/marimo"
chmod +x "$VENV/.venv/bin/marimo"
```

Alternatively, just run main marimo `pixi run marimo` and `pup.fetch` inside the notebook.  This will work the same in marimo and jupyter - no need to mess with kernels.  You can fetch more than one venv or build reproducible environments on the fly:
`pup.fetch("anotherenv", "awswrangler", "duckdb")`

6. Start marimo

```bash
newenv/.venv/bin/marimo --edit --no-token
```

Output:
```
[2024-11-27 15:24:40] fetched packages from 'newenv': /home/a/puppy-marimo/newenv/.venv/lib/python3.12/site-packages added to `sys.path`
[2024-11-27 15:24:40] 🐶 heard: pup list newenv
{
  "newenv": [
    "tldraw>=3.0.0"
  ]
}
Update available 0.9.26 → 0.9.27
Run pip install --upgrade marimo to upgrade.


        Create or edit notebooks in your browser 📝

        URL: http://localhost:2718
```

7.  Marimo runs and knows how to find tldraw:

![image](https://github.com/user-attachments/assets/daec1295-ead6-482b-adca-a679244b95c4)

**Done drawing? try another one** 

8. Create another environment with another widget:
```bash
VENV=newenv2
pup add $VENV drawdata pandas
head -5 .pixi/envs/default/bin/marimo > "$VENV/.venv/bin/marimo"
echo -e '\nimport pup\npup.fetch("'$VENV'")\n' >> "$VENV/.venv/bin/marimo"
tail -5 .pixi/envs/default/bin/marimo >> "$VENV/.venv/bin/marimo"
chmod +x "$VENV/.venv/bin/marimo"
```

9. Run marimo:
```bash
newenv2/.venv/bin/marimo --edit --no-token
```

Now [drawdata](https://github.com/koaning/drawdata) venv is active, and, as expected, tldraw is not there.

![image](https://github.com/user-attachments/assets/3a38524f-5116-43d1-a30a-3058283095f4)

**Comment by liquidcarbon** [comment]
Patching Jupyter executable in this manner is not that simple, so it's much easier to use `pup.fetch` inside the notebook.

Start using `pixi run jupyter lab` or `pixi run jupyter notebook`.

![image](https://github.com/user-attachments/assets/1cce7c21-ec12-4382-ae63-da322b361318)

**Comment by liquidcarbon** [comment]
Here's some more hackery - using `runpy`, we inject custom code prior to launching marimo - in this case, we `pup.fetch` another environment.  This works in Dockerfiles ([example on HuggingFace](https://huggingface.co/spaces/liquidcarbon/puppy-hf-marimo/blob/main/Dockerfile))

```
pixi run python -c \
    'import pup, runpy; pup.fetch("sql_and_plots"); runpy.run_path(".pixi/envs/default/bin/marimo", run_name="__main__")' \
    tutorial sql --host 0.0.0.0 --port 7860 --no-token
```

![image](https://github.com/user-attachments/assets/dd913c2c-bd20-4988-834d-b07c558e9f02)

**Comment by yharby** [comment]
I like it very cool
